crates/red_knot_python_semantic/src/types.rs

```diff
@@ -306,3 +312,359 @@ pub struct IntersectionType<'db> {
     // the intersection type does not include any value in any of these types
     negative: FxOrderSet<Type<'db>>,
 }
+
+#[allow(unused)]
+#[derive(Clone)]
+struct IntersectionTypeBuilder<'db> {
+    intersections: Vec<InnerIntersectionTypeBuilder<'db>>,
+    db: &'db dyn Db,
+}
+
+impl<'db> IntersectionTypeBuilder<'db> {
+    #[allow(dead_code)]
+    fn new(db: &'db dyn Db) -> Self {
+        Self {
+            db,
+            intersections: vec![InnerIntersectionTypeBuilder::new()],
+        }
+    }
+
+    fn empty(db: &'db dyn Db) -> Self {
+        Self {
+            db,
+            intersections: vec![],
+        }
+    }
+
+    #[allow(dead_code)]
+    fn add_positive(mut self, ty: Type<'db>) -> Self {
+        if let Type::Union(union) = ty {
+            union
+                .elements(self.db)
+                .iter()
+                .map(|elem| self.clone().add_positive(*elem))
+                .fold(
+                    IntersectionTypeBuilder::empty(self.db),
+                    |mut builder, sub| {
+                        builder.intersections.extend(sub.intersections);
+                        builder
+                    },
+                )
```

Comment:
hmm, I find the `.fold()` here quite hard to grok. I suppose it's just personal preference, but I'd find this easier to parse:

```suggestion
            let mut builder = IntersectionTypeBuilder::empty(self.db);
            for element in union.elements(self.db) {
                builder
                    .intersections
                    .extend(self.clone().add_positive(element).intersections);
            }
            builder
```

I _think_ it's equivalent? Your tests seem to pass, anyway :-)

(Same comment applies for your `add_negative()` method below)

Reply:
Some added docs here would also be great. Even after rewriting it in imperative style, it took me a _little_ bit of time to figure out what was going on here. I believe you're applying the transformation where `A & (B | C)` -> `(A & B) | (A & C)`. Correct?

Reply:
I think it is preference, and I prefer the fold :) I'll add comments to see if I can make it clearer, though.